### PR TITLE
Make ArbitraryFileIdManager filesystem-agnostic and fix Windows CI

### PR DESCRIPTION
**Pull request by kevin-bates**
The pull request addresses the issues encountered during CI on Windows platforms. There still appears to be an issue with symlinks, but only with Python 3.7 (and Windows), so that test has been skipped in that scenario (especially since support for 3.7 is going away).  I suspect this has something to do with `os.path.realpath` - which may be problematic on Windows based on this [link](https://stackoverflow.com/questions/43333640/python-os-path-realpath-for-symlink-in-windows).

A majority, nearly all, of the changes were related to how paths are normalized prior to persistence.  The `AbritraryFileIdManager` has been made filesystem-agnostic with this PR.  All slashes are converted to forward slashes with consecutive slashes removed.  As a result, even the value of `root_dir` (i.e., `content_root`) is converted, although the drive letter information remains.  This means that those using the `ArbitraryFileIdManager` with `FileContentsManagers` must be on homogeneous file systems (namely Windows vs. Unix) due to differing formats in their `root_dir` traits.

Because path normalization (and denormalization) varies between implementations, those internal methods were also abstracted.  Additionally, the `_move_recursive()`,  `_copy_recursive()` and `_delete_recursive()` methods on the ABC were updated to take a "path manager" (e.g., `os.path` or `posixpath`) since path joins are problematic when attempting to be filesystem-agnostic.

This pull request also removes a redundant select statement when checking timestamps since both the `crtime` and `mtime` values can be captured when retrieving the `path` and `ino` (inode) columns in `LocalFileIdManager.get_path()`.

I apologize for a large number of commits, but not having a Windows machine to debug directly against leads one down a long and arduous path.

Resolves: #6

### CODE REVIEW

**Comment by dlqqq** [comment]
@kevin-bates Thank you so, so much for this PR!!! I had to be pulled aside to work on Scheduler this week, so this would not have been possible without your help.

> All slashes are converted to forward slashes with consecutive slashes removed. As a result, even the value of root_dir (i.e., content_root) is converted, although the drive letter information remains.

Yup, this is fine, since the ArbitraryFileIdManager doesn't access the local filesystem.

> This means that those using the ArbitraryFileIdManager with FileContentsManagers must be on homogeneous file systems (namely Windows vs. Unix) due to differing formats in their root_dir traits.

Could you elaborate more on this point? Are you saying that an Arbitrary FIDM on Linux won't work with a FileContentsManager on Windows? It's not clear to me why this would be the case, given that the FileContentsManager always gives us a relative path delimited by forward slashes. The contents root is only used to make sure FIDM works agnostic of contents root.

File: jupyter_server_fileid/manager.py [hunk 1]
```diff
@@ -254,6 +242,45 @@ def __init__(self, *args, **kwargs):
...
+        """Accepts an API path and returns a "persistable" path, i.e. one prefixed
+        by root_dir that can then be persisted in a format relative to the implementation."""
+        # use commonprefix instead of commonpath, since root_dir may not be a
+        # absolute POSIX path.
+
+        norm_root_dir = self._normalize_separators(self.root_dir)
+        if posixpath.commonprefix([norm_root_dir, path]) != norm_root_dir:
+            path = posixpath.join(norm_root_dir, path)
+
+        path = self._normalize_separators(path)
```
Comment (dlqqq) [inline jupyter_server_fileid/manager.py:262]: I think this method would be more clear if you moved this line before the `if` block above.

File: jupyter_server_fileid/manager.py [hunk 2]
```diff
@@ -254,6 +242,45 @@ def __init__(self, *args, **kwargs):
...
+            return None
+
+        # Convert root_dir to an api path, since that's essentially what we persist.
+        norm_root_dir = self._normalize_separators(self.root_dir)
+        if posixpath.commonprefix([norm_root_dir, path]) != norm_root_dir:
+            return None
+
+        relpath = path
+        if path.startswith(norm_root_dir):
+            relpath = path[len(norm_root_dir) + 1 :]
```
Comment (dlqqq) [inline jupyter_server_fileid/manager.py:280]: 1. The `path` will always start with `norm_root_dir`, so a second conditional check isn't necessary.
2. Why not just use `posixpath.relpath()` here?

File: jupyter_server_fileid/manager.py [hunk 3]
```diff
@@ -395,12 +422,35 @@ def __init__(self, *args, **kwargs):
...
 
     def _normalize_path(self, path):
-        path = super()._normalize_path(path)
+        """Accepts an API path and returns a filesystem path, i.e. one prefixed
+        by root_dir and uses os.path.sep."""
+        # use commonprefix instead of commonpath, since root_dir may not be a
+        # absolute POSIX path.
+        if os.path.commonprefix([self.root_dir, path]) != self.root_dir:
+            path = os.path.join(self.root_dir, path)
+
```
Comment (dlqqq) [inline jupyter_server_fileid/manager.py:431]: You can remove the comment and use `os.path.commonpath()` here since LocalFileIdManager only deals with local filesystem paths.

File: jupyter_server_fileid/manager.py [hunk 4]
```diff
@@ -699,39 +740,33 @@ def get_path(self, id):
...
+            ):
+                # if file already exists at path and the ino and timestamps match,
+                # then return the correct path immediately (best case)
+                return self._from_normalized_path(path)
+
+            # otherwise, try again after calling _sync_all() to sync the Files table to the file tree
+            if retry:
+                self._sync_all()
+        else:
             return None
```
Comment (dlqqq) [inline jupyter_server_fileid/manager.py:768]: I don't think the `else` block is necessary given that we're not breaking from the loop.

File: tests/test_manager.py [hunk 5]
```diff
@@ -75,7 +110,25 @@ def get_path_nosync(fid_manager, id):
...
+
+    When testing instances of ArbitraryFileIdManager, we need to normalize the
+    path, regardless of OS, when comparing results of get_path() since this fileID
+    manager must be filesystem agnostic.
+    """
+    if isinstance(fid_manager, LocalFileIdManager):
+        path = os.path.normcase(path)
+
+    parts = path.strip("\\").split("\\")
+    return "/".join(parts)
```
Comment (dlqqq) [inline tests/test_manager.py:131]: I don't quite understand why this is necessary. Here me out:

1. The paths we use in the fixtures are always lower case and relative. So `normcase` shouldn't be doing anything.
2. We're now always using `posixpath.join()` in the fixtures, which uses `/` as the separator. So lines 130-131 shouldn't be doing anything either.

I think the changes you made to the fixtures are sufficient. Can we try removing this helper and see if CI still passes for Windows?

**Review by dlqqq — CHANGES_REQUESTED** [review]

**Comment by kevin-bates** [comment]
>> This means that those using the ArbitraryFileIdManager with FileContentsManagers must be on homogeneous file systems (namely Windows vs. Unix) due to differing formats in their root_dir traits.

> Could you elaborate more on this point? Are you saying that an Arbitrary FIDM on Linux won't work with a FileContentsManager on Windows? It's not clear to me why this would be the case, given that the FileContentsManager always gives us a relative path delimited by forward slashes. The contents root is only used to make sure FIDM works agnostic of contents root.

My use case is both contrived and hypothetical, but _if_ you had a multi-user scenario where some users were using Windows and others were using any Unix variant.  The `root_dir` for Windows might look something like `C:\\shared\directory` while on Unix, the `root_dir` would be `/shared/directory`.

All that I'm saying is that paths persisted by the Windows user using AFIDMgr look like `C:/shared/directory/path/to/myfile` when working with api-path `path/to/myfile`, while paths persisted by the Unix user look like `/shared/directory/path/to/myfile` when working with the same api-path `path/to/myfile`.  As a result, users on Unix will have different IDs for the same physical file because the `root_dir` formats are different.

So, _if_ users want to use the AFIDMgr with filesystem-based `ContentsManagers`, their users should be on homogeneous filesystem types so their `root_dir` formats are the same.

File: tests/test_manager.py (hunk 5 above)
Comment (kevin-bates) [reply tests/test_manager.py:131]: > 1. The paths we use in the fixtures are always lower case and relative. So normcase shouldn't be doing anything.

The test "paths" are indeed.  The test `root_dir` values are not.   They are represented by the pytest temporary directories and, on Windows at least, contain some uppercase characters - which `normcase()` lowers on Windows.

> 2. We're now always using posixpath.join() in the fixtures, which uses / as the separator. So lines 130-131 shouldn't be doing anything either.

True, but even `posixpath.join()` won't convert _embedded_ backslashes within one of its arguments.  e.g., `posixpath.join("c:\\path\\to\\temp\\folder", "my/api/path")` yields `c:\\path\\to\\temp\\folder/my/api/path`.

The above yields: `c:/path/to/temp/folder/my/api/path`

File: jupyter_server_fileid/manager.py (hunk 4 above)
Comment (kevin-bates) [reply jupyter_server_fileid/manager.py:768]: True.  I'll remove it.

File: jupyter_server_fileid/manager.py (hunk 1 above)
Comment (kevin-bates) [reply jupyter_server_fileid/manager.py:262]: ok.

File: jupyter_server_fileid/manager.py (hunk 3 above)
Comment (kevin-bates) [reply jupyter_server_fileid/manager.py:431]: Turns out changing this to `os.path.commonpath()` causes failures with `ValueError: Can't mix absolute and relative paths` when the `root_dir` is: '/private/var/folders/t0/lmgg5t1j35dcftfz88jf1mqw0000gn/T/pytest-of-kbates/pytest-114/test_delete_recursive_local_0/root_dir' and the `path` is 'test_path'.

Since this change isn't really germane to this PR (this code was already this way, just located in the superclass that LFIDMgr calls), I'll leave this alone due to all the failures it introduces - even on MacOS tests.

Comment (dlqqq) [reply jupyter_server_fileid/manager.py]: Bizarre! Not sure why that is; we were using `os.path.commonpath()` in the previous implementation.

File: tests/test_manager.py (hunk 5 above)
Comment (dlqqq) [reply tests/test_manager.py:131]: > The test root_dir values are not. They are represented by the pytest temporary directories and, on Windows at least, contain some uppercase characters - which normcase() lowers on Windows.

Hmm, can we just normalize the root directory argument in the `*fid_manager` fixtures in `pytest_plugin.py` instead then?

I just don't want to force usage of `normalize_path()` whenever comparing paths in the unit tests. This hinders readability and can cause loss of valuable developer time if they forget to call `normalize_path()` before assertion.

> even posixpath.join() won't convert embedded backslashes within one of its arguments. e.g., posixpath.join("c:\\path\\to\\temp\\folder", "my/api/path") yields c:\\path\\to\\temp\\folder/my/api/path.

Right, but this function will always receive a test path, which is guaranteed to be relative and generated via `posixpath.join()`. Can you try removing the backslash => forwards slash logic and see if it still works?

**Comment by dlqqq** [comment]
@kevin-bates 

> All that I'm saying is that paths persisted by the Windows user using AFIDMgr look like C:/shared/directory/path/to/myfile when working with api-path path/to/myfile, while paths persisted by the Unix user look like /shared/directory/path/to/myfile when working with the same api-path path/to/myfile. As a result, users on Unix will have different IDs for the same physical file because the root_dir formats are different.

This edge case doesn't necessarily happen for multi-user setups. If different users talk to the same file ID manager, that manager is only running on one operating system at a time, so everybody is happy. This edge case really only happens for **multiple servers running on different operating systems** all while being backed by the same local filesystem. I think excluding support for this extreme edge case is perfectly reasonable.

Thanks for addressing my review comments. I still had one concern I left open regarding the unit tests; wanted to see if we could avoid calling `normalize_path()` in the assertions by modifying the fixtures a bit more.

**Comment by dlqqq** [comment]
Plus keep in mind when using two File ID managers concurrently:

1. They should not be sharing the same DB file. Ever.
2. There's absolutely no relationship between the IDs they return. They absolutely can return different IDs for the same file.

**Comment by kevin-bates** [comment]
Yeah, I agree that the use-case is extreme.

> Plus keep in mind when using two File ID managers concurrently:
>
>1. They should not be sharing the same DB file. Ever.

I'm not sure I agree with this.  Why can't you have two FIDs running concurrently against the same database?  Yes, use of the `LocalFileIdManager` would require that both instances be on the same filesystem, but things should work.  This is a more practical use case for non-filesystem-based content.  If I use database storage across multiple servers, why can't I have multiple AFIDs sharing a database? (SQLite database issues aside.)

>2. There's absolutely no relationship between the IDs they return. They absolutely can return different IDs for the same file.

I agree.  The scope of the ID is per FID db.

> I still had one concern I left open regarding the unit tests; wanted to see if we could avoid calling normalize_path() in the assertions by modifying the fixtures a bit more.

I'll spend some time with this shortly.

**Comment by dlqqq** [comment]
> I'm not sure I agree with this. Why can't you have two FIDs running concurrently against the same database? Yes, use of the LocalFileIdManager would require that both instances be on the same filesystem, but things should work. This is a more practical use case for non-filesystem-based content. If I use database storage across multiple servers, why can't I have multiple AFIDs sharing a database? (SQLite database issues aside.)

@kevin-bates I choose my words carefully; I said **should not**, not **must not**. Meaning that this is a use-case we don't currently support and should be avoided. I'm not saying that this use case is guaranteed to fail 100% of the time. Our lack of support for this use-case is evident from the fact we're using SQLite in the current implementation. 

You're right in saying that having two File ID managers that don't use SQLite will probably work if they share a database file, either local or remote. However, it's also very unclear as to what use-case this could serve **well**. If your storage is split up across different servers, IMO that should be handled by a custom contents manager, rather than having a separate File ID manager for each server.

File: tests/test_manager.py (hunk 5 above)
Comment (kevin-bates) [reply tests/test_manager.py:131]: > Right, but this function will always receive a test path, which is guaranteed to be relative and generated via posixpath.join(). Can you try removing the backslash => forwards slash logic and see if it still works?

Again, it's the `root_dir` values that are the issue.  When they contain multiple "directories" _within_ a parameter of `join()`, those separators will not be converted.  Your idea of incorporating this in the fixture is reasonable and I'll take a look.

Another approach may be to massage `root_dir` within its validator such that it's always in a form that will be persisted.  

One of the primary problems I ran into is that the tests start with filesystem-path files but need to compare against, normalized versions of those (i.e., the "api path"), and since the test parameters are both FileSystem and NonFileSystem implementations and *nix/Windows, comparing inputs that are solely filesystem-derived requires a conversion.

I'll revisit this though - thanks.

**Comment by kevin-bates** [comment]
>> They should not be sharing the same DB file. Ever.

> I choose my words carefully; I said should not, not must not.

Fair enough. Please note that you followed up **should not** with the single-word sentence **Ever.** so I took that as a bit stronger than _should not_.

Moving on...

FWIW, here's a test failure after "removing" `normalize_path` (I updated it to simply return the `path` argument to prevent having to update all locations).  This was after massaging the `root_dir` into its persisted form when the trait is set.  Since this failure is only occurring on 4 tests for Windows, I will likely remove the `normalize_path()` call from the other locations.  The problem is that the LFID tests need the file to exist, while the AFID tests do not, but its tough to make both happy.

```python
    @pytest.mark.parametrize("stub_stat_crtime", [True, False], indirect=["stub_stat_crtime"])
    def test_get_path_oob_move_nested(fid_manager, old_path, new_path, stub_stat_crtime, fs_helpers):
        old_test_path = "test_path"
        new_test_path = os.path.join(new_path, "test_path")
        fs_helpers.touch(old_test_path)
        fid_manager.index(old_path)
        id = fid_manager.index(old_test_path)
    
        fs_helpers.move(old_path, new_path)
        fs_helpers.move(old_test_path, new_test_path)
    
>       assert fid_manager.get_path(id) == normalize_path(fid_manager, new_test_path)
E       AssertionError: assert 'new_path/test_path' == 'new_path\\test_path'
E         - new_path\test_path
E         ?         ^
E         + new_path/test_path
E         ?         ^

tests\test_manager.py:426: AssertionError
```

**Review by dlqqq — APPROVED** [review]
@kevin-bates Awesome work! Thank you so much for tackling this. I owe you an ice cream at JupyterCon if we're both able to attend. 😁
